students/ThanhNhan/lesson09/test_mr_5.py

```diff
@@ -0,0 +1,177 @@
+import os
+import pytest
+from .mail5 import Donor, DonorCollection
+
+donor_db = {"Tom Cruise": [233.23, 12.17],
+            "Alyssa Nhan": [23., 23],
+            "Jessy Apple, IIS": [33.0],
+            "Jin Unknown": [12.0, 12.0]
+            }
+
+# Test object of the name
+def test_donor_name():
+    donor = Donor("Test ASdf")
+    assert donor.name == "Test ASdf"
+
+# Count how many time donation
+def test_add_donation():
+    donor = Donor("Fred Test")
+    donor.add_donations(500)
+    assert donor.num_donations == 1
+    donor.add_donations(320)
+    assert donor.num_donations == 2
+
+#Test dollar amount input
+def test_ck_char_input():
+    donor = Donor("Fred Test")
+  
+    with pytest.raises(ValueError):
+        donor.add_donations(0.001)
+
+    with pytest.raises(ValueError):
+        donor.add_donations("asd")
+
+    with pytest.raises(ValueError):
+        donor.add_donations(-50)
+
+    with pytest.raises(ValueError):
+        donor.add_donations(0.0)
+
+
+# Validate Name Input
+def test_validate_name():
+    name1 = Donor(" ")
+    assert name1.validate_name() is False
+
+    name2 = Donor("23")
+    assert name2.validate_name() is False
+
+    name3 = Donor("asd2a adf3")
+    assert name3.validate_name() is False
+
+    name4 = Donor("3ass3 12sdf")
+    assert name4.validate_name() is False
+
+    name5 = Donor("Tom Asd")
+    assert name5.validate_name() is True
+
+    name6 = Donor("Tom")
+    assert name6.validate_name() is False
+
+    name7 = Donor("Tom ")
+    assert name7.validate_name() is False
+
+    name8 = Donor("Tom Asd III")
+    assert name8.validate_name() is True
+
+    name9 = Donor("Tom Asdasfasf, III")
+    assert name9.validate_name() is True
+
+    name10 = Donor("Tom Asdasfasf ")
+    assert name10.validate_name() is True
+
+
+def test_donor_thank_you_Letter():
+    donor = Donor("Tom Tom")
+    donor.add_donations(2432)
+    assert donor.generate_letter(donor.donations) == """
+            Dear {},
+            Thank you  for your very kind donation of ${:.2f}
+            It will be put to very good use.
+
+                            Sincerely,
+                            - The Team   
+	        """.format("Tom Tom", 2432.00)
+
+
+
+def test_DonorCollection():
+    donor = DonorCollection(donor_db)
+    assert donor.data_collection == {"Tom Cruise": [233.23, 12.17],
+                                    "Alyssa Nhan": [23., 23],
+                                    "Jessy Apple, IIS": [33.0],
+                                    "Jin Unknown": [12.0, 12.0]
+                                    }
+    
+def test_exist_list():
+    donor = DonorCollection(donor_db)
+    assert donor.exist_list() == ["Tom Cruise", "Alyssa Nhan", "Jessy Apple, IIS", "Jin Unknown"]
+
+def test_lookup_name():
+    donor = DonorCollection(donor_db, "Test As")
+    assert donor.find_donor() is False
+    donor = DonorCollection(donor_db, "Tom Cruise")
+    assert donor.find_donor() is True
```

Comment:
assert is looking for Truthiness already, so just:

assert donor.find_donor() is fine.

But this is a very odd API -- if you have the donor object, why do you need to find it??

I htikn this comes from Donor subclassing DonorCollection -- but you really don't want that.

And DonorCollection needs a way to add a donor one by one.

find_donor() should be a method of DonorCollection:

dc = DonorCollection(donor_db)
dc.add_donor("Tom Cruise")

assert dc.find_donor("Tom Cruise")